### PR TITLE
Reorder fix ttm/mod 3d arrays to be consistent with fix ttm and fix ttm/grid

### PR DESCRIPTION
**Pull request by sjplimp**
**Summary**

The 3d arrays in fix ttm/mod are indexed as array[ix][iy][iz].  The 3d arrays in fix ttm and fix ttm/grid are indexed array[iz][iy][ix].
This PR inverts the 3 indices for all the 3d arrays in fix ttm/mod to be consistent with fix ttm and ttm/grid, so that the
x values in the grid data structure vary fastest, then y, and z slowest.

**Related Issue(s)**

N/A

**Author(s)**

Steve

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

There is a small difference in the last couple digits of log file outputs in examples/ttm for in.ttm.mod.  Not sure yet
whether this is to be expected or is an indication of a bug in the refactored code.

**Implementation Notes**

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by sjplimp** [comment]
@akohlmey As indicated in the description, there is a small (last couple digits) in the logfile for the examples/ttm/in.ttm.mod script (1 and 4 procs) due to this change.  I can't see why this should be, though possibly it is simply round-off.  But I've marked this work-in-process until we figure it out.

**Comment by sjplimp** [comment]
It is more dramatically different when the grid is not 10x10x10 but 6x8x10 - so something is wrong I think.

File: src/EXTRA-FIX/fix_ttm_mod.cpp [hunk 1]
```diff
@@ -348,14 +335,14 @@ void FixTTMMod::post_force(int /*vflag*/)
...
-          double T_i = T_electron[ix][iy][iz];
-          double T_ir = T_electron[right_x][iy][iz];
-          double T_iu = T_electron[ix][right_y][iz];
-          double T_if = T_electron[ix][iy][right_z];
-          double C_i = el_properties(T_electron[ix][iy][iz]).el_heat_capacity;
-          double C_ir = el_properties(T_electron[right_x][iy][iz]).el_heat_capacity;
-          double C_iu = el_properties(T_electron[ix][right_y][iz]).el_heat_capacity;
-          double C_if = el_properties(T_electron[ix][iy][right_z]).el_heat_capacity;
+          double T_i = T_electron[iz][iy][ix];
+          double T_ir = T_electron[right_z][iy][ix];
```
Comment (akohlmey) [inline src/EXTRA-FIX/fix_ttm_mod.cpp]: @sjplimp are you sure that these changes are correct? `T_ir` changes from "right_x" to "right_z" and `T_if` the other way around.

File: src/EXTRA-FIX/fix_ttm_mod.cpp [hunk 2]
```diff
@@ -821,50 +804,50 @@ void FixTTMMod::end_of_step()
...
               if (ix >= t_surface_l) mult_factor = (intensity/(dx*skin_layer_d))*exp((-1.0)*(ix_d - surface_d)/skin_layer_d);
             }
-            if (ix < t_surface_l) net_energy_transfer_all[ix][iy][iz] = 0.0;
+            if (ix < t_surface_l) net_energy_transfer_all[iz][iy][ix] = 0.0;
             double cr_vac = 1;
-            if (T_electron_old[ix][iy][iz] == 0) cr_vac = 0;
+            if (T_electron_old[iz][iy][ix] == 0) cr_vac = 0;
             double cr_v_l_x = 1;
-            if (T_electron_old[left_x][iy][iz] == 0) cr_v_l_x = 0;
+            if (T_electron_old[left_z][iy][ix] == 0) cr_v_l_x = 0;
```
Comment (akohlmey) [inline src/EXTRA-FIX/fix_ttm_mod.cpp]: @sjplimp here is another similar case. `cr_v_l_x` is set based on `left_z` instead of `left_x` and so on.

**Comment by sjplimp** [comment]
@akohlmey Good catch on the left/right indices.  I fixed that and now the old/new answers agree even with a non cubic grid.
However the issue with the last couple digits of thermo output being different between new and old versions is still there.
For cubic or non-cubic grids.  This is the last couple digits of 16-digit precision, for 1000 step runs, 1 proc or 4 procs.

Any ideas?  I'm not seeing how operations would be performed in different orders due to the changes I made.

**Comment by akohlmey** [comment]
> Any ideas? I'm not seeing how operations would be performed in different orders due to the changes I made.

The changes could impact the loop unrolling done by the compiler at high optimization level.
You would have to compile the before and after versions with -O0 to check if that makes a difference.

**Comment by sjplimp** [comment]
@akohlmey I tried the test you suggested with -O0, comparing old and new versions.  It didn't change anything.
O0 and O3 are the same for the old version.  Also the same for the new version.  But old vs new is slightly different in
the last one of two digits of the fix output (out of 16 printed digits).  Even on timesteps 1 or 2.  Same issue with a 10^3 grid
or 6x8x10 grid.

I tried running for 10,000 steps instead of 1000 steps.  Same issue, but the differences do not grow in size.  Also,
the dynamics of the system (temp, toteng) are identical on every step.  It's only the fix output (2 values) for electron
eng and temp, where the tiny differences are, even though it's coupled to the dynamics.

The code computing these two values is quite complex in ttm/mod (more so than the other two variants), so I'm
not optimistic I can find what is causing such a tiny difference.  I cannot see any mistakes I've made with swapping
the order of indices.

Any suggestions?

**Comment by akohlmey** [comment]
> Any ideas? I'm not seeing how operations would be performed in different orders due to the changes I made.

There is at least one place, where you *did* change the order of operations:  `FixTTMMod::compute_vector()`.
I've switched back the order in which the two properties are summed and that recovers the exact output from before the other changes in this pull request.

**Comment by sjplimp** [comment]
ah, good catch.  That's a summation over the grid only for purposes of output.  Which explains why it didn't affect the dynamics, even over long timesscales.  Just the last couple digits of hi-precision values printed on the thermo line.

The way I have it (z outer, x inner) is the way it should be in compute_vector() (and everywhere else), to access the 3d arrays with stride one.  So we should leave it that way and re-bless the output.  Do you want me to produce new log files for the PR ?

**Comment by akohlmey** [comment]
I didn't see any significant performance difference either way. Just the fluctuations you see with any calculation these days in times of turbo boost and stack randomization.

**Comment by sjplimp** [comment]
@akohlmey created new log files consistent with z,y,z loop ordering - this should be ready to merge
thanks for your help

**Review by akohlmey — APPROVED** [review]

**Pull request merged**